cobalt/js_profiler/profiler.cc

```diff
@@ -0,0 +1,147 @@
+
+#include "cobalt/js_profiler/profiler.h"
+
+#include <iostream>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "cobalt/base/polymorphic_downcast.h"
+#include "cobalt/js_profiler/profiler_trace_builder.h"
+#include "cobalt/js_profiler/profiler_trace_wrapper.h"
+#include "cobalt/web/cache_utils.h"
+#include "cobalt/web/context.h"
+#include "cobalt/web/dom_exception.h"
+#include "cobalt/web/environment_settings.h"
+#include "cobalt/web/environment_settings_helper.h"
+
+namespace cobalt {
+namespace js_profiler {
+
+volatile uint32_t s_lastProfileId = 0;
+
+static constexpr int kBaseSampleIntervalMs = 10;
+
+static v8::Local<v8::String> toV8String(v8::Isolate* isolate,
+                                        const std::string& string) {
+  if (string.empty()) return v8::String::Empty(isolate);
+  return v8::String::NewFromUtf8(isolate, string.c_str(),
+                                 v8::NewStringType::kNormal, string.length())
+      .ToLocalChecked();
+}
+
+Profiler::Profiler(script::EnvironmentSettings* settings,
+                   ProfilerInitOptions options,
+                   script::ExceptionState* exception_state)
+    : cobalt::web::EventTarget(settings),
+      stopped_(false),
+      time_origin_{base::TimeTicks::Now()} {
+  profiler_id_ = nextProfileId();
+
+  const base::TimeDelta sample_interval =
+      base::Milliseconds(options.sample_interval());
+
+  int64_t sample_interval_us = sample_interval.InMicroseconds();
+
+  if (sample_interval_us < 0 ||
+      sample_interval_us > std::numeric_limits<int>::max()) {
+    sample_interval_us = 0;
+  }
+
+  int effective_sample_interval_ms =
+      static_cast<int>(sample_interval.InMilliseconds());
+  if (effective_sample_interval_ms % kBaseSampleIntervalMs != 0 ||
+      effective_sample_interval_ms == 0) {
+    effective_sample_interval_ms +=
+        (kBaseSampleIntervalMs -
+         effective_sample_interval_ms % kBaseSampleIntervalMs);
+  }
+  sample_interval_ = effective_sample_interval_ms;
+
+  auto isolate = web::get_isolate(settings);
+
+  auto status = ImplProfilingStart(
+      profiler_id_,
+      v8::CpuProfilingOptions(v8::kLeafNodeLineNumbers,
+                              options.max_buffer_size(), sample_interval_us),
+      settings);
+
+  if (status == v8::CpuProfilingStatus::kAlreadyStarted) {
+    web::DOMException::Raise(web::DOMException::kInvalidStateErr,
+                             "Profiler Already started", exception_state);
+  } else if (status == v8::CpuProfilingStatus::kErrorTooManyProfilers) {
+    web::DOMException::Raise(web::DOMException::kInvalidStateErr,
+                             "Too Many Profilers", exception_state);
+  }
+}
+
+v8::CpuProfilingStatus Profiler::ImplProfilingStart(
+    std::string profiler_id, v8::CpuProfilingOptions options,
+    script::EnvironmentSettings* settings) {
+  auto isolate = web::get_isolate(settings);
+  cpu_profiler_ = v8::CpuProfiler::New(isolate);
```

Comment:
Given that `cpu_profiler_` is  a bare pointer, doesn't that leak this v8 profiler instance when this JS object is destroyed, or even when profiling is stopped?

Separately, it seems that if this Profiler JS object is ever destroyed before/without calling `Profiler.stop()` (e.g. when the JS object's last reference is removed and the GC runs and deletes it) and waiting for the promise to resolve, the profiling is not actually  stopped, which may result in callbacks to the `ProfilerMaxSamplesDelegate`, which (if the null check is added) shouldn't result in crashes, but will result in unnecessary CPU usage since profiling will continue indefinitely.

Reply:
Unfortunately I can't seem to make a unique_ptr from an object with a private destructor (at least not without a custom Deleter). So I added a destructor to the Profiler class, which should be called by JS Garbage collection when appropriate. Please let me know if this is sufficient, thank you.